core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,449 @@
+                        break;
+                    } else {
+                        added.add(tombstone);
+                    }
+                }
+                // look through the front, starting at the beginning, for removed tombstones
+                final Tombstone firstTombstone = currentTombstones.getFirst();
+                for (Iterator<Tombstone> iter = previousTombstones.iterator(); iter.hasNext();) {
+                    final Tombstone tombstone = iter.next();
+                    if (firstTombstone.equals(tombstone)) {
+                        // current list has this entry, so no more removing
+                        break;
+                    } else {
+                        removed.add(tombstone);
+                    }
+                }
+            }
+            Collections.reverse(added); // added list was created from back to front, so have to reverse
+            this.added = Collections.unmodifiableList(added);
+            this.removed = Collections.unmodifiableList(removed);
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeList(added);
+            out.writeList(removed);
+        }
+
+        @Override
+        public IndexGraveyard apply(final IndexGraveyard old) {
+            final Deque<Tombstone> oldQueue = old.tombstones;
+            final Deque<Tombstone> newQueue = new ArrayDeque<>(oldQueue);
+            for (Tombstone tombstone : removed) {
+                newQueue.remove(tombstone);
+            }
+            for (Tombstone tombstone : added) {
+                newQueue.addLast(tombstone);
+            }
+            return new IndexGraveyard(newQueue);
+        }
+
+        /** The index tombstones (deletions) that were added between two states */
+        public List<Tombstone> getAdded() {
+            return added;
+        }
+
+        /** The index tombstones (deletions) that were removed between two states */
+        public List<Tombstone> getRemoved() {
+            return removed;
+        }
+    }
+
+    /**
+     * An individual tombstone entry for representing a deleted index.
+     */
+    final public static class Tombstone implements ToXContent, StreamInputReader<Tombstone>, Writeable<Tombstone> {
+
+        public static final Tombstone PROTOTYPE = new Tombstone(new Index("_na_", "_na_"), 0L, 0L);
+        private static final String INDEX_UUID_KEY = "index_uuid";
+        private static final String INDEX_NAME_KEY = "index_name";
+        private static final String DELETE_DATE_KEY = "delete_date";
+        private static final String DELETE_DATE_NS_KEY = "delete_date_ns";
+        static final ObjectParser<Tombstone.Builder, Void> TOMBSTONE_PARSER = new ObjectParser<>("tombstoneEntry");
+        static {
+            TOMBSTONE_PARSER.declareString(Tombstone.Builder::indexUUID, new ParseField(INDEX_UUID_KEY));
```

Comment:
can we add the serialization logic we need to the Index object it self? we're likely to use it in other places.